### ISSUE TITLE
didUpdateWidget() does not take into account changes in navbar items

### ISSUE THREAD

**Issue by celariss** [maheshmnj/navbar_router#38]
**Describe the bug**
When the NavBarRouter object is given new items during build process (after a setstate() call for example), _NavbarRouterState does not rebuild its keys and items members, resulting in showing incorrect text.

**To Reproduce**
In my case, since I implemented a dynamic language change in my whole app, the NavBarRouter ctor in my RootPageState build method is given new text that is ignored.

**Expected behavior**
NavBarRouter should always reset its items and keys members in didUpdateWidget() without condition to take into account any change in widget items.

I made a "local" fix by commenting a few lines in _NavbarRouterState.didUpdateWidget() (navbar_router.dart) :
```dart
void didUpdateWidget(covariant NavbarRouter oldWidget) {
    /// update animation
    if (widget.destinationAnimationCurve !=
            oldWidget.destinationAnimationCurve ||
        widget.destinationAnimationDuration !=
            oldWidget.destinationAnimationDuration) {
      initAnimation();
    }
    // HERE IS THE FIX (no condition for items/keys reset)
    //if (widget.destinations.length != oldWidget.destinations.length ||
    //        widget.type != oldWidget.type) {
          NavbarNotifier.length = widget.destinations.length;
          clearInitialization();
          initialize();
    //}
    super.didUpdateWidget(oldWidget);
  }
```

Thank you for your work :)

**Comment by maheshmnj** [comment]
Hi @celariss, Thanks for filing the issue. We should definitely fix that, Please feel free to make a PR if you are interested and I would be happy to review it.

**Comment by maheshmnj** [comment]
Closing as fixed in [v0.7.1](https://pub.dev/packages/navbar_router/changelog)

**Comment by maheshmnj** [comment]
Reopening this is still reproducible as of v0.7.3, 
reproduction: updating selectedIcon does not rebuild navbar

**Comment by maheshmnj** [comment]
Updating the priority this is really a show blocker for production

**Comment by bebaoboy** [comment]
I also faced the same issue. I was using a 3rd theme switcher package with circular animation (not the original fade animation of flutter when going from light to dark theme) and a button to switch theme mode dynamically, directly placed on the appbar so I can see the theme changing on the current screen.
When I press the button the theme change; however, the icon and background color on the navbar stay the same 😂 tried setState, but still. 
So my work around for this is just rebuild the entire navbar and its destinations by setting the navbar items inside the build() method of my screen containing the navbar 😭

**Comment by maheshmnj** [comment]
The root cause of this issue is one liner perhaps heres the minimal dartpad reproducible this code sample should output false but it returns true.

```
import 'package:flutter/material.dart';

class NavbarItem {
  const NavbarItem(this.iconData, this.text,
      {this.backgroundColor, this.child, this.selectedIcon});

  /// IconData for the navbar item
  final IconData iconData;

  /// label for the navbar item
  final String text;

  /// background color for the navbar item when type is [NavbarType.shifting]
  /// ignored otherwise
  final Color? backgroundColor;

  /// When child is specified, the item will be rendered as a floating navbar
  /// ignoring iconData and text
  final Widget? child;

  /// Widget to show when the item is selected
  final Widget? selectedIcon;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is NavbarItem &&
          runtimeType == other.runtimeType &&
          iconData == other.iconData &&
          text == other.text &&
          child.runtimeType == other.child.runtimeType &&
          selectedIcon.runtimeType == other.selectedIcon.runtimeType &&
          backgroundColor == other.backgroundColor;

  @override
  int get hashCode =>
      iconData.hashCode ^
      text.hashCode ^
      child.hashCode ^
      selectedIcon.hashCode ^
      backgroundColor.hashCode;
}

int main(){
  
  const nav1 = NavbarItem(Icons.person_outline, 'Me',
          backgroundColor:Colors.blue,
          selectedIcon: const Icon(
            Icons.person,
            color: Colors.red,
            size: 26,
          ));
  
    const nav2 = NavbarItem(Icons.person_outline, 'Me',
          backgroundColor:Colors.blue,
          selectedIcon: const Icon(
            Icons.person,
            size: 26,
          ));
  
  print(nav1 == nav2);
  
  return 0;
}
```

**Comment by maheshmnj** [comment]
This should be fixed now, Updates to change in selectedIcon widget type and  size, icon and color of icon are taken into consideration.

**Comment by bebaoboy** [comment]
Wow, the fix was incredibly simple! But sometimes the selectedIcon is not Icon and maybe user uses a custom child widget. Could the test be passed in this situation? 


Or should we force user to give some kind of "key" for those widgets?

**Comment by maheshmnj** [comment]
Yes you are right, I think user will need to pass a key for the custom key, I am not really clear on that part th current fix only considers Icon being passed to selectedIcon. Your inputs are welcome on the right approch to follow in this case.

**Comment by maheshmnj** [comment]
Released in [v0.7.4](https://pub.dev/packages/navbar_router/changelog)
